source/Halibut/Transport/Streams/WriteIntoMemoryBufferStream.cs

```diff
@@ -150,9 +150,10 @@ public async Task WriteBufferToUnderlyingStream(CancellationToken cancellationTo
             if (usingMemoryBuffer)
             {
                 memoryBuffer.Position = 0;
-                await memoryBuffer.CopyToAsync(innerStream, 8192, cancellationToken);
-
+                // Mark we are not using the memory buffer here, since if the write below fails
```

Comment:
This makes sense, and will work in this case 😄 

But some follow-on thoughts:
- I think this issue can happen during a `Write` and `WriteAsync` operation as well. Maybe we need to fix those two also?
- Using `usingMemoryBuffer` works fine, and I'm ok if we want to keep using it. But it feels a little like it's serving 2 purposes:
 - Knowing which stream to write to when writing
 - Choosing to write on dispose
- Do you think introducing a new variable for writing on dispose would make things more clear? Or just be more noise 🤔

Reply:
Probably more noise I think.

Reply:
Did you want the other methods order changed?

Reply:
I think it's worthwhile. You found this issue for this particular code path, and I think it can happen the other way as well.

Since we're going to the effort to fix this one, I'd do the complete set.

You can verify by writing the tests, and then fixing it 😁